plugins/providers/shield/resource.go

```diff
@@ -0,0 +1,157 @@
+package shield
+
+import (
+	"fmt"
+
+	"github.com/odpf/guardian/domain"
+)
+
+const (
+	ResourceTypeTeam         = "team"
+	ResourceTypeProject      = "project"
+	ResourceTypeOrganization = "organization"
+)
+
+type Metadata struct {
+	Email   string `json:"email" mapstructure:"email"`
+	Privacy string `json:"privacy" mapstructure:"privacy"`
+	Slack   string `json:"slack" mapstructure:"slack"`
+}
+
+type User struct {
+	ID    string `json:"id" mapstructure:"id"`
+	Name  string `json:"name" mapstructure:"name"`
+	Email string `json:"email" mapstructure:"email"`
+}
+
+type Team struct {
+	ID       string   `json:"id" mapstructure:"id"`
+	Name     string   `json:"name" mapstructure:"name"`
+	Slug     string   `json:"slug" mapstructure:"slug"`
+	OrgId    string   `json:"orgId" mapstructure:"orgId"`
+	Metadata Metadata `json:"metadata" mapstructure:"metadata"`
+	Admins   []string `json:"admin" mapstructure:"admin"`
+}
+
+type Project struct {
+	ID     string   `json:"id" mapstructure:"id"`
+	Name   string   `json:"name" mapstructure:"name"`
+	Slug   string   `json:"slug" mapstructure:"slug"`
+	OrgId  string   `json:"orgId" mapstructure:"orgId"`
+	Admins []string `json:"admin" mapstructure:"admin"`
+}
+
+type Organization struct {
+	ID     string   `json:"id" mapstructure:"id"`
+	Name   string   `json:"name" mapstructure:"name"`
+	Slug   string   `json:"slug" mapstructure:"slug"`
+	Admins []string `json:"admin" mapstructure:"admin"`
+}
+
+func (t *Team) FromDomain(r *domain.Resource) error {
+	if r.Type != ResourceTypeTeam {
+		return ErrInvalidResourceType
+	}
+
+	resourceDetails := r.Details
+	t.ID = resourceDetails["id"].(string)
+	t.OrgId = resourceDetails["orgId"].(string)
+	t.Name = r.Name
+
+	if resourceDetails["admins"] == nil {
+		t.Admins = []string{}
+	} else {
+		adminsInterface := resourceDetails["admins"].([]interface{})
+		admins := make([]string, len(adminsInterface))
+		for i, v := range adminsInterface {
+			admins[i] = v.(string)
+		}
+		t.Admins = admins
+	}
+
+	metadataInterface := resourceDetails["metadata"].(interface{})
+	metadata, ok := metadataInterface.(Metadata)
+	if ok {
+		t.Metadata = metadata
+	}
+
+	return nil
+}
+
+func (t *Team) ToDomain() *domain.Resource {
+	return &domain.Resource{
+		Type: ResourceTypeTeam,
+		Name: t.Name,
+		URN:  fmt.Sprintf("team:%v", t.ID),
+		Details: map[string]interface{}{
```

Comment:
Is there a specific reason for map[string]interface{} here compared to a concrete struct for the details field? All the details in this file can be modeled into a struct...

Reply:
Details is the field in Resource struct which common for all resource types.
Initialising the details object here.